libs/community/langchain_community/vectorstores/cloudSQL.py

```diff
@@ -0,0 +1,475 @@
+import uuid
+
+from typing import Any, Callable, Dict, List, Optional, Tuple, Type
+
+from langchain_core.documents import Document
+from langchain_core.vectorstores import VectorStore
+from langchain_core.embeddings import Embeddings
+from langchain_community.vectorstores.utils import maximal_marginal_relevance
+
+from google.cloud.sql.connector import Connector
+import google.auth
+from google.auth.transport.requests import Request
+from google.auth.compute_engine import _metadata
+
+from sqlalchemy import inspect
+from sqlalchemy.dialects.postgresql import JSON, UUID
+
+from pgvector.sqlalchemy import Vector
+
+def _get_IAM_user(credentials):
+    """Get user/service account name"""
+
+    # Checks for service account
+    if hasattr(credentials, "service_account_email"):
+        return credentials.service_account_email[:-20]
+    else:
+        # Defaults to getting the user account
+        command = ["gcloud", "auth", "list", "--filter=status:ACTIVE", "--format=value(account)"]
+        result = subprocess.run(command, stdout=subprocess.PIPE, text=True)
+        if result.returncode == 0:
+            return result.stdout.strip()
+        else:
+            return f"Error: {result.stderr}"
```

Comment:
I assume this is trying to detect the user from the IAM credentials? Ideally we should not be using gcloud, and should stick to the standard Application Default Credentials here. 

You should be able to get the email more consistently by calling the oauth2 api. It should look something like this, but in async:
```python
# make GET request
url = f"https://oauth2.googleapis.com/tokeninfo?access_token={credentials.token}"
response = await client_session.get(url)
# check if email is present in response
if "email" in response:
    # format_user formats database username based on database type
    user = format_user(response["email"], database_type)
```

Reply:
What is the database_type here? Can we just slice the response with condition check for 'iam'  in the response["email"]

Reply:
Ah this is because MySQL and Postgres have different requirements for IAM name -- it can't just be [-20]. I think its for MySQL we need to remove anything after the "@" and for postgres we remove ".gserviceaccount.com"